cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/DiskRemovalGoal.java

```diff
@@ -0,0 +1,169 @@
+import java.util.List;
+import java.util.ArrayList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.ACCEPT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.GoalUtils.MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING;
+
+
+/**
+ * Soft goal to move the replicas to different log dir.
+ */
+public class DiskRemovalGoal implements Goal {
+    private static final Logger LOG = LoggerFactory.getLogger(DiskRemovalGoal.class);
+    private final ProvisionResponse _provisionResponse;
+
+    protected final Map<Integer, Set<String>> _brokerIdAndLogdirs;
+
+    protected final double _errorMargin;
+
+    public DiskRemovalGoal(Map<Integer, Set<String>> brokerIdAndLogdirs, double errorMargin) {
+        _provisionResponse = new ProvisionResponse(ProvisionStatus.UNDECIDED);
+        _brokerIdAndLogdirs = brokerIdAndLogdirs;
+        _errorMargin = errorMargin;
+    }
+
+    private void sanityCheckOptimizationOptions(OptimizationOptions optimizationOptions) {
+        if (optimizationOptions.isTriggeredByGoalViolation()) {
+            throw new IllegalArgumentException(String.format("%s goal does not support use by goal violation detector.", name()));
+        }
+    }
+
+    @Override
+    public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, OptimizationOptions optimizationOptions) {
+        sanityCheckOptimizationOptions(optimizationOptions);
+
+        for (Map.Entry<Integer, Set<String>> brokerIdLogDirs : _brokerIdAndLogdirs.entrySet()) {
+            Integer brokerId = brokerIdLogDirs.getKey();
+            Set<String> logDirsToRemove = brokerIdLogDirs.getValue();
+            relocateBrokerLogDirs(clusterModel, brokerId, logDirsToRemove);
+        }
+
+        return true;
+    }
+
+    /**
+     * This method relocates the replicas on the provided log dirs to other log dirs of the same broker.
+     * @param clusterModel the cluster model
+     * @param brokerId the id of the broker where the movement will take place
+     * @param logDirsToRemove the set of log dirs to be removed from the broker
+     */
+    private void relocateBrokerLogDirs(ClusterModel clusterModel, Integer brokerId, Set<String> logDirsToRemove) {
+        Broker currentBroker = clusterModel.broker(brokerId);
+        List<Disk> remainingDisks = new ArrayList<>();
+        currentBroker.disks().stream().filter(disk -> !logDirsToRemove.contains(disk.logDir())).forEach(remainingDisks::add);
+
+        for (String logDirToRemove : logDirsToRemove) {
+            Set<Replica> replicasToMove = currentBroker.disk(logDirToRemove).replicas();
+            for (int i = 0; i < replicasToMove.size(); i++) {
+                Replica replica = replicasToMove.iterator().next();
+                relocateReplicaIfPossible(clusterModel, brokerId, remainingDisks, replica);
+                LOG.info(String.format("Could not move replica %s to any of the remaining disks.", replica));
```

Comment:
SLF4J loggers support placeholders for strings with the following syntax:
```suggestion
                LOG.info("Could not move replica {} to any of the remaining disks.", replica);
```

Reply:
That's cool! Didn't know about it. Removed call to string format.